bin/deepstate/main_manticore.py

```diff
@@ -382,6 +382,7 @@ def run_tests(args, state, apis):
   #pool = multiprocessing.Pool(processes=max(1, args.num_workers))
   results = []
   mc = DeepManticore(state)
+  mc.context['apis'] = apis
```

Comment:
Should fix the `KeyError`, since the new `DeepManticore` instance never calls `read_api_table` to instantiate `self.context['apis']`.

While this does allow the executor to continue symex, one thing to note (not sure if related) is that with the error fix, the executor fails when attempting to remove the initial state from the workspace:

```
ERROR:deepstate.mcore:Uncaught exception: <class 'FileNotFoundError'>
Traceback (most recent call last):
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/deepstate-0.1-py3.6.egg/deepstate/main_manticore.py", line 374, in run_test
    do_run_test(state, apis, test, hook_test)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/deepstate-0.1-py3.6.egg/deepstate/main_manticore.py", line 334, in do_run_test
    m = manticore.native.Manticore(state, sys.argv[1:])
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/manticore.py", line 42, in __init__
    super().__init__(initial_state, workspace_url=workspace_url, policy=policy, **kwargs)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/manticore.py", line 1004, in __init__
    super().__init__(*args, **kwargs)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/manticore.py", line 285, in __init__
    self._put_state(initial_state)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/manticore.py", line 430, in _put_state
    state_id = self._save(state, state_id=state.id)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/manticore.py", line 382, in _save
    state._id = self._workspace.save_state(state, state_id=state_id)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/workspace.py", line 477, in save_state
    self.rm_state(state_id)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/workspace.py", line 488, in rm_state
    return self._store.rm(f"{self._prefix}{state_id:08x}{self._suffix}")
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/workspace.py", line 293, in rm
    os.remove(path)
FileNotFoundError: [Errno 2] No such file or directory: '/home/vagrant/deepstate/build/mcore_0xbk7a_j/state_00000000.pkl'
```

Reply:
If you patch manticore, you get into this error:

```
$ deepstate-manticore examples/Crash 
2019-06-25 11:50:35,314: [5637] m.n.manticore:INFO: Loading program examples/Crash
2019-06-25 11:50:37,410: [5637] m.n.manticore:INFO: Loading program examples/Crash
INFO:deepstate.mcore:Running 1 tests across 1 workers
2019-06-25 11:50:41,421: [5693] m.c.manticore:INFO: Deprecated!
2019-06-25 11:50:41,443: [5693] m.n.manticore:INFO: Loading program examples/Crash
2019-06-25 11:50:43,145: [5741] m.c.worker:ERROR: Exception in state 0: KeyError('stream_6',)
Traceback (most recent call last):
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/core/worker.py", line 119, in run
    current_state.execute()
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/state.py", line 20, in execute
    result = self._platform.execute()
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/platforms/linux.py", line 2390, in execute
    self.current.execute()
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/cpu/abstractcpu.py", line 962, in execute
    self._publish("will_execute_instruction", self.PC, insn)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/utils/event.py", line 128, in _publish
    self._publish_impl(_name, *args, **kwargs)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/utils/event.py", line 145, in _publish_impl
    sink._publish_impl(_name, *args, **kwargs)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/utils/event.py", line 145, in _publish_impl
    sink._publish_impl(_name, *args, **kwargs)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/utils/event.py", line 143, in _publish_impl
    sink._publish_impl(_name, self, *args, **kwargs)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/utils/event.py", line 136, in _publish_impl
    callback(robj(), *args, **kwargs)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/manticore.py", line 260, in _hook_callback
    cb(state)
  File "/home/g/.local/lib/python3.6/site-packages/deepstate-0.1-py3.6.egg/deepstate/main_manticore.py", line 261, in <lambda>
    return lambda state: state.invoke_model(func)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/state.py", line 65, in invoke_model
    self._platform.invoke_model(model, prefix_args=(self,))
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/platforms/platform.py", line 53, in invoke_model
    self._function_abi.invoke(model, prefix_args)
  File "/home/g/.local/lib/python3.6/site-packages/manticore-0.3.0-py3.6.egg/manticore/native/cpu/abstractcpu.py", line 368, in invoke
    result = model(*arguments)
  File "/home/g/.local/lib/python3.6/site-packages/deepstate-0.1-py3.6.egg/deepstate/main_manticore.py", line 196, in hook_LogStream
    DeepManticore(state).api_log_stream(level)
  File "/home/g/.local/lib/python3.6/site-packages/deepstate-0.1-py3.6.egg/deepstate/common.py", line 637, in api_log_stream
    stream = self.context[stream_id]
KeyError: 'stream_6'
```

Reply:
The race condition (`FileNotFound` exception) in accessing the mcore workspace seems to be caused by the fact that with each `deepstate-manticore` run, two `Manticore` instances are spawned off, each possibly attempting to read/write from the same. The `KeyError` may also be a result of that, as the populated context manager from `DeepManticore.begin_test` isn't present once we attach hooks to various `DeepState_*` functions.

One way I tried to mitigate these exceptions would be changing 

```suggestion
463   base = get_base(m)
464   apis = mc.read_api_table(ea_of_api_table, base)
465   del mc
466
467   m.add_hook(setup_ea, lambda state: run_tests(args, state, apis))
468   m.run()
```

to 

```suggestion
464   base = get_base(m)
465   mc.context['apis'] = mc.read_api_table(ea_of_api_table, base)
466   tests = mc.find_test_cases()
467   del mc
468
469   L.info("Running {} tests across {} workers".format(
470     len(tests), args.num_workers))
471
472   for test in tests:
473     m.add_hook(setup_ea, lambda state: run_test(state, apis, test))
474
475   #m.add_hook(setup_ea, lambda state: run_tests(args, state, apis))
476   m.run()
```
While this rids both exceptions, and allows execution until it reaches a `SyscallNotImplemented` instruction:

```
Traceback (most recent call last):
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/core/worker.py", line 119, in run
    current_state.execute()
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/native/state.py", line 20, in execute
    result = self._platform.execute()
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/platforms/linux.py", line 2397, in execute
    self.syscall()
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/platforms/linux.py", line 2224, in syscall
    return self._syscall_abi.invoke(implementation)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/native/cpu/abstractcpu.py", line 426, in invoke
    ret = super().invoke(model, prefix_args)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/native/cpu/abstractcpu.py", line 368, in invoke
    result = model(*arguments)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/platforms/platform.py", line 22, in unimplemented
    return wrapped(*args, **kwargs)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/platforms/linux_syscall_stubs.py", line 1178, in sys_wait4
    return self.complicated_success(61)
  File "/home/vagrant/deepstate/venv/lib/python3.6/site-packages/manticore/platforms/linux_syscall_stubs.py", line 36, in complicated_success
    raise SyscallNotImplemented(num, amd64[num])
manticore.platforms.platform.SyscallNotImplemented: Syscall index "61" (sys_wait4) not implemented.
```

Reply:
I believe this is caused by manticore executing instructions after the symbolic hook (which should never happen, since after hook symbolic engine should run). Maybe @pgoodman can confirm this?